### DISCUSSION TITLE
Set parent component state from nested component data

### DISCUSSION THREAD

**Discussion by believelody** [alpinejs/alpine#4136] [1. Help] [answered]
Hi guys, I have a `<ul>` component with its children `<li>` component. A short example will be like so:
```
<ul x-data="{ itemsToDelete: new Set([]) }" . . .>
    <template x-for="item in itemsToDelete">
        <li x-data="{ id: item.id }" . . .>
            <span x-text="item.name" />
            <button @click="itemsToDelete.include(id) ? itemsToDelete.delete(id) : itemsToDelete.add(id)">Add/Delete</button>
        </li>
    </template>
</ul>
```
In case above, mutate itemsToDelete in `<li>` has effect to `<ul>`.
I want to register each component to its own function data, for instance Alpine.data("itemList", itemListData) for `<ul>` and Alpine.data("item", itemData) for` <li>`.

How can I access and mutate reactively "itemsToDelete" defined in itemListData from itemData like I do in the templating case ?

Sincerely

**Comment by SimoTod** [comment]
Child components can access the parent scope. Whether you use Alpine.data or put your data directly in x-data won't change the feature.

your example has some issues though:
- x-for needs an array, it can't iterate over Sets
- you can use a self closing tag with span, only void element can be expressed in that way (https://developer.mozilla.org/en-US/docs/Glossary/Void_element) but it's not really necessary in modern html
- you need to pass the whole element, not the id to the Set operations (and add will never happen)

**Reply by believelody** [reply]
Hi, thanks for your reply and your clarification points. My bad, I wanted to simplify. I messed up things. I write below an improved version of my example:
```
     <ul x-data={`{ items: [ { id: 1, name: "item-1" }, { id: 2, name: "item-2" }, { id: 3, name: "item-3" }, { id: 4, name: "item-4" } ], itemsToDelete: new Set([]) }`} x-effect="console.log(itemsToDelete.size)">
        <template x-for="item in items">
          <li x-data="{ id: item.id, name: item.name }" class="flex items-center gap-x-4">
            <span x-text="name" />
            <button x-show="!itemsToDelete.has(id)" x-on:click="itemsToDelete.add(id)">
              Add
            </button>
            <button x-show="itemsToDelete.has(id)" x-on:click="itemsToDelete.delete(id)">
              Remove
            </button>
          </li>
        </template>
      </ul>
```
So, we iterate over items and the purpose is to fill "itemsToDelete". If you can try and tell me if it's works for you.
I will ask my question in next comment.

**Reply by SimoTod** [reply]
Still half broken (wrong/missing quotes, duplicate parenthesis, span tag still using a not allowed syntax.
It has to be something like https://codepen.io/SimoTod/pen/gOyeQoy

**Reply by believelody** [reply]
Now I want to separate things into two functions data and register them with Alpine.data. For example:
```
Alpine.data("list", function() {
    return {
        items: [ { id: 1, name: "item-1" }, { id: 2, name: "item-2" }, { id: 3, name: "item-3" }, { id: 4, name: "item-4" } ],
        itemsToDelete: new Set([]),
    }
});

Alpine.data("item", function(id, name) {
    return {
        id,
        name,
        isInItemsToDelete() {
            // How can I access "itemsToDelete" ? Will it be reactive ?
        },
        add() {
            // How can I mutate "itemsToDelete" here ?
        },
        delete() {
            // How can I mutate "itemsToDelete" here ?
        },
    }
});
```

By doing so, I can have in my html something like this:
```
<ul x-data={"list()"} x-effect="console.log(itemsToDelete.size)">
        <template x-for="item in items">
          <li x-data="item(item.id, item.name)" class="flex items-center gap-x-4">
            <span x-text="name"></span>
            <button x-show="!isInItemsToDelete()" x-on:click="add()">
              Add
            </button>
            <button x-show="isInItemsToDelete()" x-on:click="delete()">
              Remove
            </button>
          </li>
        </template>
      </ul>
```

**Reply by believelody** [reply]
Hi @SimoTod , ooh, I'm using vs code and JSX as template engine. I understand now why some things broke on your side. I saw your codePen, thank you very much. Just below your last reply I write what I want to do. Thanks for your time.

**Reply by SimoTod** [reply, marked as answer]
right, when you post on a public forum, you should try to write plain html otherwise people cannot copy paste your examples and they spend additional time trying to fix the errors.
You can use this.itemsToDelete in the child Alpine.data (and yes, it will be reactive, you should really try it first before asking the question :D ) although from a software design perspective is odd. 
I would just add the functions to manipulate itemsToDelete in the parent component and use them in the child html.

Try a few options and see what you like best

**Reply by believelody** [reply]
Ooh, I had no clue parent data can be used this way 😱  ! I even did it in a js file and it works like a charm 🤯 
Thanks a lot for your guidance.
This is the final result:

In html file
```
<ul x-data="items" x-effect="console.log(itemsToDelete.size)">
  <template x-for="item in items">
    <li x-data="singleitem(item)" class="flex items-center gap-x-4">
      <span x-text="name"></span>
      <button x-show="!isInItemsToDelete()" x-on:click="add()">
        Add
      </button>
      <button x-show="isInItemsToDelete()" x-on:click="remove()">
        Remove
      </button>
    </li>
  </template>
  <li>debug '<span x-text="Array.from(itemsToDelete)"></span>'</li>
</ul> 
```

In js file
```
Alpine.data("items", () => ({
  items: [
    { id: 1, name: "item-1" },
    { id: 2, name: "item-2" },
    { id: 3, name: "item-3" },
    { id: 4, name: "item-4" }
  ],
  itemsToDelete: new Set([])
}));

Alpine.data("singleitem",(item) => ({
  id: item.id,
  name: item.name,
  add() {
    this.itemsToDelete.add(this.id);
  },
  remove() {
    this.itemsToDelete.delete(this.id);
  },
  isInItemsToDelete() {
    return this.itemsToDelete.has(this.id);
  }
}));
```
I let the code pen link for those who will need it: https://codepen.io/believelody/pen/mdgxoWz?editors=1011
